torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -234,8 +236,64 @@ inline void errorIfCapturingNonCapturableNCCL() {
   }
 }
 
+void markRankNotStuck(char *rankStatusHelper, int rank) {
```

Comment:
I think there is a problem here where we are not talking into account the async nature of cuda. For example the user could schedule multiple collectives like this:

```
w1 = allreduce(...)
w2 = allgather(...)
w3 = reduce_scatter(...)
w1.wait()
w2.wait()
w3.wait()
```

Now the first 3 lines would set the status to `0x01` and then `w1.wait()` would set status to `0x00` after finishing. However the allgather could get stuck and as a result `w2.wait()` gets stuck, but our status will still be `0x00`.

I think we need to write additional details to the file:

1. The status (which we already have here).
2. The sequence number of the collective (can be retrieved from: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp#L434)
3. The type of the collective: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp#L32

This way, we can then analyze the file and sequence numbers to accurately determine which rank was stuck on which collective.

Reply:
This does however make the file management a lot trickier. We no longer have one entry per rank, but multiple entries (for all inflight collectives) per rank and there is not necessarily a fixed upper bound here. In practice there would probably be a handful of collectives in flight, but theoretically you can schedule hundreds of collectives together.

Reply:
Is knowing what collective got stuck necessary? If not, what if `+1` upon entering the collective and `-1` upon exit. Would that work? Also, do we need to synchronize access for this read/write?

With the current scheme of one byte per rank, that would allow up to 255 inflight collectives.  If a higher number is required, changing to multiple bytes instead of one or even a file/rank should be easy.

Reply:
> Is knowing what collective got stuck necessary?

This is incredibly helpful and maybe could be thought of as a hard requirement for this effort in my mind (doesn't mean it needs to be in the first version of this soln, but eventually). This is because the debugger is usually workflow author/application developer who we want to give as much information to as possible. Having the specific collective the rank got stuck on, collectives other ranks were doing, and as much of stacktrace of possible has been historically valuable and led to quicker debug.